Dcc

stefan11111 opened the pull request.

emekoi commented on kirc.c:
```diff
@@ -916,36 +998,113 @@ static void handle_user_input(state l)
+	    if(*(tok+strlen(tok)+1))
+		*(tok+strlen(tok))=' ';
+	    raw("privmsg %s :%s\r\n", l->buf + 5, tok);
+	    if(strncmp(l->buf + 5, "NickServ", 8))
```
why is `NickServ` special-cased here?

stefan11111 replied:
Because you usually send out passwords to NickServ, and you don't want to print the password back to the screen in case someone sees it. Also NickServ responds with good messages so you don't need confirmation of sending something.

emekoi commented:
@stefan11111 i'll try to code up a fix for the lint issue from codacy and push it to #132. if you have trouble rebasing off of that i can try either pushing to [@stefan11111/dcc](https://github.com/stefan11111/kirc) or just posting a text patch here. i'll also try adding file resume support.

stefan11111 commented:
Push the changes to @stefan11111/dcc if you can. I will see if I can get codacity to stop complaining about that if statement, though I don't think it is actually an issue.

stefan11111 commented:
@emekoi 
Is there any way to resolve the last codacity warning?

stefan11111 commented:
@mcpcpc  @emekoi 
I have written a patch that de-nests almost all of the functions, except the ones related to @emekoi's patch. This is done by adding proper return statements to all of the unnecessarily nested void functions. I will post the patch in a separate comment as it is quite long. I am planning to add it to this pr.

I have found while testing that @emekoi's dcc resume patch will make kirc error out if I download a file to completion, try to download it again and have kirc send an XDCC CANCEL request, delete it with rm -rf and try to download it again, kirc errors out with the following error message:
`open: No such file or directory`

stefan11111 commented:
Here is the patch

```
diff --git a/kirc.c b/kirc.c
index cafca55..4cff9d7 100644
--- a/kirc.c
+++ b/kirc.c
@@ -110,8 +110,7 @@ static int get_columns(int ifd, int ofd)
     }
 }
 
-static void
-buffer_position_move(state l, ssize_t dest, ssize_t src, size_t size)
+static void buffer_position_move(state l, ssize_t dest, ssize_t src, size_t size)
 {
     memmove(l->buf + l->posb + dest, l->buf + l->posb + src, size);
 }
@@ -283,80 +282,87 @@ static int edit_insert(state l, char *c)
                 if (write(STDOUT_FILENO, c, clenb) == -1) {
                     return -1;
                 }
-            } else {
-                refresh_line(l);
+                return 0;
             }
-        } else {
-            buffer_position_move_end(l, clenb, 0);
-            memmove(l->buf + l->posb, c, clenb);
-            l->posu8++;
-            l->lenu8++;
-            l->posb += clenb;
-            l->lenb += clenb;
             refresh_line(l);
+            return 0;
         }
+        buffer_position_move_end(l, clenb, 0);
+        memmove(l->buf + l->posb, c, clenb);
+        l->posu8++;
+        l->lenu8++;
+        l->posb += clenb;
+        l->lenb += clenb;
+        refresh_line(l);
     }
     return 0;
 }
 
 static void edit_move_left(state l)
 {
-    if (l->posb > 0) {
-        l->posb = u8_previous(l->buf, l->posb);
-        l->posu8--;
-        refresh_line(l);
+    if (l->posb <= 0){
+        return;
     }
+    l->posb = u8_previous(l->buf, l->posb);
+    l->posu8--;
+    refresh_line(l);
+    return;
 }
 
 static void edit_move_right(state l)
 {
-    if (l->posu8 != l->lenu8) {
-        l->posb = u8_next(l->buf, l->posb);
-        l->posu8++;
-        refresh_line(l);
+    if (l->posu8 == l->lenu8) {
+        return;
     }
+    l->posb = u8_next(l->buf, l->posb);
+    l->posu8++;
+    refresh_line(l);
 }
 
 static void edit_move_home(state l)
 {
-    if (l->posb != 0) {
-        l->posb = 0;
-        l->posu8 = 0;
-        refresh_line(l);
+    if (l->posb == 0) {
+        return;
     }
+    l->posb = 0;
+    l->posu8 = 0;
+    refresh_line(l);
 }
 
 static void edit_move_end(state l)
 {
-    if (l->posu8 != l->lenu8) {
-        l->posb = l->lenb;
-        l->posu8 = l->lenu8;
-        refresh_line(l);
+    if (l->posu8 == l->lenu8) {
+        return;
     }
+    l->posb = l->lenb;
+    l->posu8 = l->lenu8;
+    refresh_line(l);
 }
 
 static void edit_delete(state l)
 {
-    if ((l->lenu8 > 0) && (l->posu8 < l->lenu8)) {
-        size_t this_size = u8_next(l->buf, l->posb) - l->posb;
-        buffer_position_move_end(l, 0, this_size);
-        l->lenb -= this_size;
-        l->lenu8--;
-        refresh_line(l);
+    if ((l->lenu8 <= 0) || (l->posu8 >= l->lenu8)) {
+        return;
     }
+    size_t this_size = u8_next(l->buf, l->posb) - l->posb;
+    buffer_position_move_end(l, 0, this_size);
+    l->lenb -= this_size;
+    l->lenu8--;
+    refresh_line(l);
 }
 
 static void edit_backspace(state l)
 {
-    if ((l->posu8 > 0) && (l->lenu8 > 0)) {
-        size_t prev_size = l->posb - u8_previous(l->buf, l->posb);
-        buffer_position_move_end(l, (ssize_t) - prev_size, 0);
-        l->posb -= prev_size;
-        l->lenb -= prev_size;
-        l->posu8--;
-        l->lenu8--;
-        refresh_line(l);
+    if ((l->posu8 <= 0) || (l->lenu8 <= 0)) {
+        return;
     }
+    size_t prev_size = l->posb - u8_previous(l->buf, l->posb);
+    buffer_position_move_end(l, (ssize_t) - prev_size, 0);
+    l->posb -= prev_size;
+    l->lenb -= prev_size;
+    l->posu8--;
+    l->lenu8--;
+    refresh_line(l);
 }
 
 static void edit_delete_previous_word(state l)
@@ -398,41 +404,43 @@ static void edit_delete_line_to_end(state l)
 
 static void edit_swap_character_w_previous(state l)
 {
-    if (l->posu8 > 0 && l->posu8 < l->lenu8) {
-        char aux[8];
-        ssize_t prev_size = l->posb - u8_previous(l->buf, l->posb);
-        ssize_t this_size = u8_next(l->buf, l->posb) - l->posb;
-        memmove(aux, l->buf + l->posb, this_size);
-        buffer_position_move(l, -prev_size + this_size, -prev_size, prev_size);
-        memmove(l->buf + l->posb - prev_size, aux, this_size);
-        if (l->posu8 != l->lenu8 - 1) {
-            l->posu8++;
-            l->posb += this_size;
-        }
-        refresh_line(l);
+    if (l->posu8 <= 0 || l->posu8 >= l->lenu8) {
+        return;
     }
+    char aux[8];
+    ssize_t prev_size = l->posb - u8_previous(l->buf, l->posb);
+    ssize_t this_size = u8_next(l->buf, l->posb) - l->posb;
+    memmove(aux, l->buf + l->posb, this_size);
+    buffer_position_move(l, -prev_size + this_size, -prev_size, prev_size);
+    memmove(l->buf + l->posb - prev_size, aux, this_size);
+    if (l->posu8 != l->lenu8 - 1) {
+        l->posu8++;
+        l->posb += this_size;
+    }
+    refresh_line(l);
 }
 
 static void edit_history(state l, int dir)
 {
-    if (history_len > 1) {
-        free(history[history_len - (1 + l->history_index)]);
-        history[history_len - (1 + l->history_index)] = strdup(l->buf);
-        l->history_index += (dir == 1) ? 1 : -1;
-        if (l->history_index < 0) {
-            l->history_index = 0;
-            return;
-        } else if (l->history_index >= history_len) {
-            l->history_index = history_len - 1;
-            return;
-        }
-        strncpy(l->buf, history[history_len - (1 + l->history_index)],
-                l->buflen);
-        l->buf[l->buflen - 1] = '\0';
-        l->lenb = l->posb = strnlen(l->buf, MSG_MAX);
-        l->lenu8 = l->posu8 = u8_length(l->buf);
-        refresh_line(l);
+    if (history_len <= 1) {
+        return;
+    }
+    free(history[history_len - (1 + l->history_index)]);
+    history[history_len - (1 + l->history_index)] = strdup(l->buf);
+    l->history_index += (dir == 1) ? 1 : -1;
+    if (l->history_index < 0) {
+        l->history_index = 0;
+        return;
+    }
+    if (l->history_index >= history_len) {
+        l->history_index = history_len - 1;
+        return;
     }
+    strncpy(l->buf, history[history_len - (1 + l->history_index)],l->buflen);
+    l->buf[l->buflen - 1] = '\0';
+    l->lenb = l->posb = strnlen(l->buf, MSG_MAX);
+    l->lenu8 = l->posu8 = u8_length(l->buf);
+    refresh_line(l);
 }
 
 static int history_add(const char *line)
@@ -486,36 +494,38 @@ static void edit_escape_sequence(state l, char seq[3])
             if ((seq[2] == '~') && (seq[1] == 3)) {
                 edit_delete(l); /* Delete key. */
             }
-        } else {
-            switch (seq[1]) {
+            return;
+        }
+        switch (seq[1]) {
             case 'A':
                 edit_history(l, 1);
-                break;          /* Up */
+                return;          /* Up */
             case 'B':
                 edit_history(l, 0);
-                break;          /* Down */
+                return;          /* Down */
             case 'C':
                 edit_move_right(l);
-                break;          /* Right */
+                return;          /* Right */
             case 'D':
                 edit_move_left(l);
-                break;          /* Left */
+                return;          /* Left */
             case 'H':
                 edit_move_home(l);
-                break;          /* Home */
+                return;          /* Home */
             case 'F':
                 edit_move_end(l);
-                break;          /* End */
-            }
+                return;;          /* End */
         }
-    } else if (seq[0] == 'O') { /* ESC O sequences. */
+        return;
+    }
+    if (seq[0] == 'O') { /* ESC O sequences. */
         switch (seq[1]) {
         case 'H':
             edit_move_home(l);
-            break;              /* Home */
+            return;              /* Home */
         case 'F':
             edit_move_end(l);
-            break;              /* End */
+            return;;              /* End */
         }
     }
 }
@@ -917,17 +927,25 @@ static void handle_ctcp(param p)
     const char *message = p->message + 1;
     if (!strncmp(message, "VERSION", 7)) {
         raw("NOTICE %s :\001VERSION kirc " VERSION "\001\r\n", p->nickname);
-    } else if (!strncmp(message, "TIME", 7)) {
+        return;
+    }
+    if (!strncmp(message, "TIME", 7)) {
         char buf[26];
         if (!ctime_now(buf)) {
             raw("NOTICE %s :\001TIME %s\001\r\n", p->nickname, buf);
         }
-    } else if (!strncmp(message, "CLIENTINFO", 10)) {
+        return;
+    }
+    if (!strncmp(message, "CLIENTINFO", 10)) {
         raw("NOTICE %s :\001CLIENTINFO " CTCP_CMDS "\001\r\n", p->nickname);
-    } else if (!strncmp(message, "PING", 4)) {
+        return;
+    }if (!strncmp(message, "PING", 4)) {
         raw("NOTICE %s :\001%s\r\n", p->nickname, message);
-    } else if (!strncmp(message, "DCC", 3)) {
+        return;
+    }
+    if (!strncmp(message, "DCC", 3)) {
         handle_dcc(p);
+        return;
     }
 }
 
@@ -1003,21 +1021,31 @@ static void raw_parser(char *string)
             raw("JOIN #%s\r\n", tok);
         }
         return;
-    } else if (!strncmp(p.command, "QUIT", 4)) {
+    }
+    if (!strncmp(p.command, "QUIT", 4)) {
         param_print_quit(&p);
-    } else if (!strncmp(p.command, "PART", 4)) {
+        printf("\x1b[0m\r\n");
+        return;
+    }if (!strncmp(p.command, "PART", 4)) {
         param_print_part(&p);
-    } else if (!strncmp(p.command, "JOIN", 4)) {
+        printf("\x1b[0m\r\n");
+        return;
+    }if (!strncmp(p.command, "JOIN", 4)) {
         param_print_join(&p);
-    } else if (!strncmp(p.command, "NICK", 4)) {
+        printf("\x1b[0m\r\n");
+        return;
+    }if (!strncmp(p.command, "NICK", 4)) {
         param_print_nick(&p);
-    } else if (!strncmp(p.command, "PRIVMSG", 7)) {
+        printf("\x1b[0m\r\n");
+        return;
+    }if (!strncmp(p.command, "PRIVMSG", 7)) {
         param_print_private(&p);
         message_wrap(&p);
-    } else {
-        param_print_channel(&p);
-        message_wrap(&p);
+        printf("\x1b[0m\r\n");
+        return;
     }
+    param_print_channel(&p);
+    message_wrap(&p);
     printf("\x1b[0m\r\n");
 }
 
@@ -1051,8 +1079,7 @@ static int handle_server_message(void)
                 message_buffer[i + 1] = '\0';
                 raw_parser(message_buffer);
                 message_buffer[i + 1] = saved_char;
-                memmove(&message_buffer, &message_buffer[i + 1],
-                        message_end - i - 1);
+                memmove(&message_buffer, &message_buffer[i + 1], message_end - i - 1);
                 message_end = message_end - i - 1;
                 i = 0;
             }

```

stefan11111 commented:
@mcpcpc  @emekoi 
Note that this patch may make codacity to scream about a lot of already merged bugs as this patch addresses most of the functions.
I have found no bugs added by this patch.
It does not change functionality, it just aligns a function like the rest of them and makes functions return early by adding proper return statements

mcpcpc commented on kirc.c:
```diff
@@ -969,8 +1200,75 @@ static void open_tty()
+        goto handle_err;
+    }
+
+    if (n) {
```
For readability, this should be an explicit comparison. i.e. `n > 0`.  this also leaves `n = 0` unchecked.

stefan11111 replied:
I'll let @emekoi make changes to this and I will merge it to dcc.
The issue I mentioned above should also be resolved.

stefan11111 replied:
@mcpcpc @emekoi 
This was fixed in my last commits.

stefan11111 commented:
@mcpcpc  @emekoi 
I have written yet another patch that fixes bugs and codacity warnings.
It also does not appear to introduce any new bugs.
I am planning to merge in into dcc.
Here is the patch:

```
diff --git a/kirc.c b/kirc.c
index c8c1393..029c8fb 100644
--- a/kirc.c
+++ b/kirc.c
@@ -5,9 +5,6 @@
 
 #include "kirc.h"
 
-#define STR_(a) #a
-#define STR(a) STR_(a)
-
 static void free_history(void)
 {
     if (history) {
@@ -101,7 +98,7 @@ static int get_columns(int ifd, int ofd)
         if (cols > start) {
             char seq[32];
             snprintf(seq, sizeof(seq), "\x1b[%dD", cols - start);
-            if (write(ofd, seq, strnlen(seq, MSG_MAX)) == -1) {
+            if (write(ofd, seq, strnlen(seq, 32)) == -1) {
             }
         }
         return cols;
@@ -251,18 +248,18 @@ static void refresh_line(state l)
         txtlenb += u8_next(buf + txtlenb, 0);
     ab_initialize(&ab);
     snprintf(seq, sizeof(seq), "\r");
-    ab_append(&ab, seq, strnlen(seq, MSG_MAX));
+    ab_append(&ab, seq, strnlen(seq, 64));
     ab_append(&ab, l->prompt, l->plenb);
     ab_append(&ab, "> ", 2);
     ab_append(&ab, buf, txtlenb);
     snprintf(seq, sizeof(seq), "\x1b[0K");
-    ab_append(&ab, seq, strnlen(seq, MSG_MAX));
+    ab_append(&ab, seq, strnlen(seq, 64));
     if (posu8 + plenu8) {
         snprintf(seq, sizeof(seq), "\r\x1b[%dC", (int)(posu8 + plenu8));
     } else {
         snprintf(seq, sizeof(seq), "\r");
     }
-    ab_append(&ab, seq, strnlen(seq, MSG_MAX));
+    ab_append(&ab, seq, strnlen(seq, 64));
     if (write(fd, ab.b, ab.len) == -1) {
     }
     ab_free(&ab);
@@ -300,7 +297,7 @@ static int edit_insert(state l, char *c)
 
 static void edit_move_left(state l)
 {
-    if (l->posb <= 0){
+    if (l->posb == 0){
         return;
     }
     l->posb = u8_previous(l->buf, l->posb);
@@ -341,7 +338,7 @@ static void edit_move_end(state l)
 
 static void edit_delete(state l)
 {
-    if ((l->lenu8 <= 0) || (l->posu8 >= l->lenu8)) {
+    if ((l->lenu8 == 0) || (l->posu8 >= l->lenu8)) {
         return;
     }
     size_t this_size = u8_next(l->buf, l->posb) - l->posb;
@@ -353,7 +350,7 @@ static void edit_delete(state l)
 
 static void edit_backspace(state l)
 {
-    if ((l->posu8 <= 0) || (l->lenu8 <= 0)) {
+    if ((l->posu8 == 0) || (l->lenu8 == 0)) {
         return;
     }
     size_t prev_size = l->posb - u8_previous(l->buf, l->posb);
@@ -404,7 +401,7 @@ static void edit_delete_line_to_end(state l)
 
 static void edit_swap_character_w_previous(state l)
 {
-    if (l->posu8 <= 0 || l->posu8 >= l->lenu8) {
+    if (l->posu8 == 0 || l->posu8 >= l->lenu8) {
         return;
     }
     char aux[8];
@@ -446,15 +443,12 @@ static void edit_history(state l, int dir)
 static int history_add(const char *line)
 {
     char *linecopy;
-    if (history_max_len == 0) {
-        return 0;
-    }
     if (history == NULL) {
-        history = malloc(sizeof(char *) * history_max_len);
+        history = malloc(sizeof(char *) * HIS_MAX);
         if (history == NULL) {
             return 0;
         }
-        memset(history, 0, (sizeof(char *) * history_max_len));
+        memset(history, 0, (sizeof(char *) * HIS_MAX));
     }
     if (history_len && !strcmp(history[history_len - 1], line)) {
         return 0;
@@ -463,9 +457,9 @@ static int history_add(const char *line)
     if (!linecopy) {
         return 0;
     }
-    if (history_len == history_max_len) {
+    if (history_len == HIS_MAX) {
         free(history[0]);
-        memmove(history, history + 1, sizeof(char *) * (history_max_len - 1));
+        memmove(history, history + 1, sizeof(char *) * (HIS_MAX - 1));
         history_len--;
     }
     history[history_len] = linecopy;
@@ -514,7 +508,7 @@ static void edit_escape_sequence(state l, char seq[3])
                 return;          /* Home */
             case 'F':
                 edit_move_end(l);
-                return;;          /* End */
+                return;          /* End */
         }
         return;
     }
@@ -525,20 +519,20 @@ static void edit_escape_sequence(state l, char seq[3])
             return;              /* Home */
         case 'F':
             edit_move_end(l);
-            return;;              /* End */
+            return;              /* End */
         }
     }
 }
 
 static int edit(state l)
 {
-    char c, seq[3];
-    int ret = 0;
+    char c;
     ssize_t nread = read(ttyinfd, &c, 1);
     if (nread <= 0) {
-        ret = 1;
-    } else {
-        switch (c) {
+        return 1;
+    }
+    char seq[3];
+    switch (c) {
         case 13:
             edit_enter();
             return 1;           /* enter */
@@ -548,71 +542,68 @@ static int edit(state l)
         case 127:              /* backspace */
         case 8:
             edit_backspace(l);
-            break;              /* ctrl-h */
+            return 0;              /* ctrl-h */
         case 2:
             edit_move_left(l);
-            break;              /* ctrl-b */
+            return 0;              /* ctrl-b */
         case 6:
             edit_move_right(l);
-            break;              /* ctrl-f */
+            return 0;              /* ctrl-f */
         case 1:
             edit_move_home(l);
-            break;              /* Ctrl+a */
+            return 0;              /* Ctrl+a */
         case 5:
             edit_move_end(l);
-            break;              /* ctrl+e */
+            return 0;              /* ctrl+e */
         case 23:
             edit_delete_previous_word(l);
-            break;              /* ctrl+w */
+            return 0;              /* ctrl+w */
         case 21:
             edit_delete_whole_line(l);
-            break;              /* Ctrl+u */
+            return 0;              /* Ctrl+u */
         case 11:
             edit_delete_line_to_end(l);
-            break;              /* Ctrl+k */
+            return 0;              /* Ctrl+k */
         case 14:
             edit_history(l, 0);
-            break;              /* Ctrl+n */
+            return 0;              /* Ctrl+n */
         case 16:
             edit_history(l, 1);
-            break;              /* Ctrl+p */
+            return 0;              /* Ctrl+p */
         case 20:
             edit_swap_character_w_previous(l);
-            break;              /* ctrl-t */
+            return 0;              /* ctrl-t */
         case 27:
             edit_escape_sequence(l, seq);
-            break;              /* escape sequence */
+            return 0;              /* escape sequence */
         case 4:                /* ctrl-d */
-            if (l->lenu8 > 0) {
-                edit_delete(l);
-            } else {
+            if (l->lenu8 == 0) {
                 history_len--;
                 free(history[history_len]);
-                ret = -1;
+                return -1;
             }
-            break;
-        default:
-            if (u8_character_start(c)) {
-                char aux[8];
-                aux[0] = c;
-                int size = u8_character_size(c);
-                for (int i = 1; i < size; i++) {
-                    nread = read(ttyinfd, aux + i, 1);
-                    if ((aux[i] & 0xC0) != 0x80) {
-                        break;
-                    }
-                }
-                aux[size] = '\0';
-                if (edit_insert(l, aux)) {
-                    ret = -1;
-                }
+            edit_delete(l);
+            return 0;
+    }
+    if (u8_character_start(c)) {
+        char aux[8];
+        aux[0] = c;
+        int size = u8_character_size(c);
+        for (int i = 1; i < size; i++) {
+            nread = read(ttyinfd, aux + i, 1);
+            if ((aux[i] & 0xC0) != 0x80) {
+                break;
             }
-            break;
+        }
+        aux[size] = '\0';
+        if (edit_insert(l, aux)) {
+            return -1;
         }
     }
-    return ret;
+    return 0;
 }
 
+
 static void state_reset(state l)
 {
     l->plenb = strnlen(l->prompt, MSG_MAX);
@@ -724,9 +715,9 @@ static void message_wrap(param p)
         return;
     }
     char *tok;
-    size_t wordwidth, spacewidth = 1;
     size_t spaceleft = p->maxcols - (p->nicklen + p->offset);
     for (tok = strtok(p->message, " "); tok != NULL; tok = strtok(NULL, " ")) {
+        size_t wordwidth, spacewidth = 1;
         wordwidth = strnlen(tok, MSG_MAX);
         if ((wordwidth + spacewidth) > spaceleft) {
             printf("\r\n%*.s%s ", (int)p->nicklen + 1, " ", tok);
@@ -1002,7 +993,6 @@ static void raw_parser(char *string)
     if (olog) {
         log_append(string, olog);
     }
-    char *tok;
     param_t p = {
         .prefix = strtok(string, " ") + 1,
         .suffix = strtok(NULL, ":"),
@@ -1016,6 +1006,7 @@ static void raw_parser(char *string)
         .offset = 0
     };
     if (!strncmp(p.command, "001", 3) && chan != NULL) {
+        char *tok;
         for (tok = strtok(chan, ",|"); tok != NULL; tok = strtok(NULL, ",|")) {
             strcpy(cdef, tok);
             raw("JOIN #%s\r\n", tok);
@@ -1201,7 +1192,7 @@ static void handle_user_input(state l)
 }
 
 static unsigned long long htonll(unsigned long long x) {
-    union { int i; char c; } u = { 1 };
+    union { char c; } u = { 1 };
     return u.c ? ((unsigned long long)htonl(x & 0xFFFFFFFF) << 32) | htonl(x >> 32) : x;
 }
 
@@ -1384,7 +1375,7 @@ int main(int argc, char **argv)
     }
     dcc_sessions.sock_fds[CON_MAX] = (struct pollfd){.fd = ttyinfd,.events = POLLIN};
     dcc_sessions.sock_fds[CON_MAX + 1] = (struct pollfd){.fd = conn,.events = POLLIN};
-    char usrin[MSG_MAX];
+    char usrin[MSG_MAX] = "";
     state_t l = {
         .buf = usrin,
         .buflen = MSG_MAX,
diff --git a/kirc.h b/kirc.h
index be767ee..e71dfb5 100644
--- a/kirc.h
+++ b/kirc.h
@@ -16,6 +16,8 @@
 #define FNM_MAX 255
 #define CON_MAX 10
 #define CBUF_SIZ 1024
+#define STR_(a) #a
+#define STR(a) STR_(a)
 
 #if defined(__FreeBSD__) || defined(__OpenBSD__)
 #include <sys/types.h>
@@ -61,7 +63,6 @@ static int ttyinfd = STDIN_FILENO;
 static struct termios orig;
 static int rawmode = 0;
 static int atexit_registered = 0;
-static int history_max_len = HIS_MAX;
 static int history_len = 0;
 static char **history = NULL;
```

stefan11111 commented:
While codacity complains about 2 new issues, these are old issues that weren't added in this patch

stefan11111 commented:
@emekoi 
check line 1232: const char *err_str = "";
It should probably be replaced.

stefan11111 commented:
@mcpcpc  @emekoi 
I have found the issue regarding the DCC resume patch.
The issue was caused by forgetting to close a file descriptor.
Also, the open_tty function seems to have no use.
Regarding the if (n) statement, I have fixed that too.
I will merge a patch that addresses all of these issues

mcpcpc commented on kirc.c:
```diff
@@ -249,18 +248,18 @@ static void refresh_line(state l)
     ab_initialize(&ab);
     snprintf(seq, sizeof(seq), "\r");
-    ab_append(&ab, seq, strnlen(seq, MSG_MAX));
+    ab_append(&ab, seq, strnlen(seq, 64));
```
why `64` for maxlen?

stefan11111 replied:
Because seq was only defined so big.
This is how it was defined before I made any changes.
Making the number any higher only introduced possible overflows.
Another solution would have been defining seq larger, but I decided to not increase it's size as to preserve original functionality.
If you want, I can increase all sizes to MSG_MAX, but it made no sense to define seq with length 64 and treat it as if it was of length MSG_MAX.

stefan11111 replied:
@mcpcpc @emekoi 
So, is everything all right? Or should we use the second approach?

mcpcpc replied:
> @mcpcpc @emekoi So, is everything all right, or should we use the second approach?

leaving it as-is should be fine.

stefan11111 commented:
@mcpcpc @emekoi 
Any more issues regarding this pr?

stefan11111 commented:
@emekoi 
This seems to work just a fine:
```
diff --git a/kirc.c b/kirc.c
index dd2750d..0dda997 100644
--- a/kirc.c
+++ b/kirc.c
@@ -1194,8 +1194,7 @@ static void handle_user_input(state l)
 }

 static unsigned long long htonll(unsigned long long x) {
-    union { char c; } u = { 1 };
-    return u.c ? ((unsigned long long)htonl(x & 0xFFFFFFFF) << 32) | htonl(x >> 32) : x;
+    return ((unsigned long long)htonl(x & 0xFFFFFFFF) << 32) | htonl(x >> 32);
 }

 static void usage(void)

```

stefan11111 commented:
@emekoi 
Do you still have any problems downloading files over 4 GiB?
I can download such files with no issue. I have teseted it on this file:
[https://nibl.co.uk/search?query=%5BUTW-THORA%5D+Evangelion+3.33+You+Can+%28Not%29+Redo+%5BBD%5D%5B1080p%2Cx264%2Cflac%5D%5BF2060CF5%5D.mkv](url)

mcpcpc commented:
> @mcpcpc @emekoi Any more issues regarding this pr?

I honestly have not had the bandwidth to really test or exercise this PR/feature. With that said, I will trust you and emekoi to make the call when this is ready to merge. Once merged, I withhold from generating a new release tarball for a few weeks in-case any obvious issues arise or if we need to roll back for any reason.

stefan11111 commented:
Damn, I always forget that not everyone has very fast internet. Guess it must suck.
Here in Romania the minimum plan is 500mbps.

mcpcpc commented:
@emekoi any other concerns with this PR before I merge?

stefan11111 commented:
Can anyone see what the difference is between `lenu8 += u8_character_start(*(s++));` and
`lenu8+=u8_character_start(*s); s++;` ?

Also, why and we not simply use `strnlen(s, MAXLEN);` or `strnlen(s, MSG_MAX)/sizeof(*s);` instead of using a loop?

stefan11111 commented:
@emekoi  @mcpcpc 
I wrote a patch that adds ipv6 support to @emekoi's DCC patch. It is quite ugly as it has up to 5(FIVE) levels of indentation, but it works. I made a pr in my repo. I am planing to merge it. Tell me if you see something off about it or about anything else in this pr.

stefan11111 commented:
@mcpcpc @emekoi 
I pushed my changes to dcc. This patch adds dcc support for ipv6 addresses. I have also written special casting for the /nick command to update the nick variable. Without it, you could not use the dcc patch after using /nick.
I also de-nested all of the code, making it more readable. Do you have anything to add to this pr?

stefan11111 commented:
@emekoi @mcpcpc 
Anyone there?

mcpcpc commented:
> @emekoi @mcpcpc
> Anyone there?

I’ve been traveling for the past two weeks.  Won’t be able to take a look until next week.

stefan11111 commented:
@mcpcpc happy vacation

stefan11111 commented:
@mcpcpc @emekoi 
I wrote a patch to add timestamps to the received messages.
https://github.com/stefan11111/kirc/pull/13

While testing it, I found some already existing bugs.
If a portion of the received message is longer the window width, it no longer wraps the message.
If a portion of the received message is exactly equal in length to the width, an extra newline is printed.
I have not fixed these as of writing this message.

My patch has the side effect of limiting nicks to 18 characters as opposed to the current 26 to make room for the timestamp. It shouldn't make much of a difference. Bots that only send server messages still have the 26 character limit as their messages aren't time-stamped. These numbers can be increased or decreased, but they have to be increased/decreased with the same value.

stefan11111 commented:
One way to solve all of these issues is to implement hexchat-text's interface.

stefan11111 commented:
@emekoi 
Simply sending an XDCC CANCEL when a file is already downloaded leads to all the running and pending downloads being canceled. This should be dealt with in a different way.

stefan11111 commented:
Anyone there?

mcpcpc commented:
> Anyone there?

Yup! This is a sizable patch/feature to review. I keep sitting down and starting to review it, only to get interrupted. I see three possible solutions to get to merge faster:

1. Maintain a separate feature branch for the time being and merge patches in until confidence in code quality can be verified. I can add a comment to the master branch for direction on this.

2. Have someone else take the time (other than me) to review this patch for code quality.

3. Merge to master and accept the risk. Continue to patch merge patches as issues arise.

Please let me know your thoughts.

stefan11111 commented:
The decision is ultimately up to you.
I would say the first option is the way to go as more people magically appearing to review this is unlikely.
I have been using this for the last few months. The only thing I've noticed it what I mentioned above regarding cancelling all the downloads when a file has been already downloaded.

stefan11111 commented:
https://github.com/stefan11111/kirc/pull/16 is a way to solve that issue.

mcpcpc commented:
> The decision is ultimately up to you.
> I would say the first option is the way to go as more people magically appearing to review this is unlikely.
> I have been using this for the last few months. The only thing I've noticed it what I mentioned above regarding cancelling all the downloads when a file has been already downloaded.

Alright. Let’s give option 1 a try. I created a new branch for merging: https://github.com/mcpcpc/kirc/tree/kirc-dcc.  Feel free to create a new PR to this one and I will merge as soon as you are ready.

mcpcpc commented:
Merged as a separate feature #135
